qiskit/transpiler/passes/synthesis/unitary_synthesis.py

```diff
@@ -442,84 +581,115 @@ def __init__(self):
         if qubits_tuple in self._decomposer_cache:
             return self._decomposer_cache[qubits_tuple]
 
-        matching = {}
-        reverse = {}
-        kak_gates = _find_matching_kak_gates(target)
-        euler_basis_gates = _find_matching_euler_bases(target)
-        decomposers_2q = []
+        # available instructions on this qubit pair, and their associated property.
+        available_2q_basis = {}
+        available_2q_props = {}
+        try:
+            keys = target.operation_names_for_qargs(qubits_tuple)
+            for key in keys:
+                available_2q_basis[key] = target.operation_from_name(key)
+                available_2q_props[key] = target[key][qubits_tuple]
+        except KeyError:
+            pass
+        try:
+            keys = target.operation_names_for_qargs(reverse_tuple)
+            for key in keys:
+                if key not in available_2q_basis:
+                    available_2q_basis[key] = target.operation_from_name(key)
+                    available_2q_props[key] = target[key][reverse_tuple]
+        except KeyError:
+            pass
+        if not available_2q_basis:
+            raise TranspilerError(
+                f"Target has no gates available on qubits {qubits} " f"to synthesize over."
+            )
+        # available decomposition basis on each of the qubits of the pair
+        # NOTE: assumes both qubits have the same single-qubit gates
+        available_1q_basis = _find_matching_euler_bases(target, qubits_tuple[0])
+
         # find all decomposers
-        for kak_gate, euler_basis in product(kak_gates, euler_basis_gates):
-            gate_name = None
-            if isinstance(kak_gate, tuple):
-                gate_name = kak_gate[1]
-                kak_gate = kak_gate[0]
-            if isinstance(kak_gate, RZXGate):
-                backup_optimizer = TwoQubitBasisDecomposer(
-                    CXGate(), euler_basis=euler_basis, pulse_optimize=pulse_optimize
-                )
-                decomposer = XXDecomposer(
-                    euler_basis=euler_basis, backup_optimizer=backup_optimizer
-                )
-                if gate_name is not None:
-                    decomposer.gate_name = gate_name
-                decomposers_2q.append(decomposer)
-            elif kak_gate is not None:
-                decomposer = TwoQubitBasisDecomposer(
-                    kak_gate, euler_basis=euler_basis, pulse_optimize=pulse_optimize
+        # TODO: reduce number of decomposers here somehow
+        decomposers = []
+
+        def is_supercontrolled(gate):
+            coords = TwoQubitWeylDecomposition(Operator(gate).data)
+            return isclose(coords.a, pi / 4) and isclose(coords.c, 0.0)
+
+        def is_controlled(gate):
+            coords = TwoQubitWeylDecomposition(Operator(gate).data)
+            return isclose(coords.b, 0.0) and isclose(coords.c, 0.0)
+
+        # possible supercontrolled decomposers (i.e. TwoQubitBasisDecomposer)
```

Comment:
for this particular class (TwoQubitBasisDecomposer) it already seems to have a `is_supercontrolled`  attribute although if it's not then your function seems necessary.

Reply:
i tried to leverage this but it's not convenient because I have to first make a whole decomposer instance for each gate that I want to check. I think this lightweight check is more straightforward.